### PR TITLE
[Bug fix] Disable memory reuse on feeded variables

### PR DESCRIPTION
**Pull request by sneaxiy**
It is better that memory reuse passes do not reuse the variable that users feeds, because users may feed them to another network, which may cause calculation errors. This PR disables memory reuse on feeded variables.

For example:

```Python
tensor = fluid.LoDTensor()
... 

# If tensor is reused by other variables, its data would be changed.
exe.run(program1, feed={'image': tensor})

# The following training may be wrong if tensor is reused in the previous network.
# However, users may not know tensor has been changed.
exe.run(program2, feed={'image': tensor})
```

Although memory reuse is disabled in feeded variables, garbage collection still works on them.

### CODE REVIEW

File: paddle/fluid/framework/ir/memory_optimize_pass/memory_optimization_var_info.h [hunk 1]
```diff
@@ -44,17 +48,30 @@ class MemOptVarInfo {
...
+
+  bool IsSkippedMemoryReuse() const {
+    return skip_memory_reuse_ || skip_all_memory_optimization_;
+  }
 
   const std::string &Name() const { return name_; }
 
  private:
   std::string name_;
   size_t ref_cnt_;
```
Comment (zhhsplendid) [inline paddle/fluid/framework/ir/memory_optimize_pass/memory_optimization_var_info.h:85]: Please add comment for the difference of these two variables.

Comment (zhhsplendid) [reply paddle/fluid/framework/ir/memory_optimize_pass/memory_optimization_var_info.h:85]: I meant, ref_cnt_ and runtime_ref_cnt_

Comment (sneaxiy) [reply paddle/fluid/framework/ir/memory_optimize_pass/memory_optimization_var_info.h:85]: Done.

Comment (zhhsplendid) [reply paddle/fluid/framework/ir/memory_optimize_pass/memory_optimization_var_info.h:85]: Great!

**Review by zhhsplendid — APPROVED** [review]
LGTM

**Pull request merged**